Drop Caml runtimes and primitives

cometkim opened the pull request:
Resolves #5779

## Rationale

This removes runtime modules and primitives that exist for OCaml compatibility.

The original BuckleScript tried distinguishing between OCaml-inherited primitives and custom primitives (`?prim`) and js primitives (`#prim`), but it didn't stick well.

Since we gave up OCaml compatibility, it added unnecessary complexity with indirection, making it hard to predict where specialization would be performed.

So here I unify all primitives into a single prefix grammar of `%`. There is still `#` primitive, but it exists temporarily for compatibility or is meant to be internal to the compiler and not exposed to the frontend.

## Changes

This PR contains many breaking changes

- Removed unused code path
- Removed unnecessary test files
- Changed runtime semantics
  - `Primitive_object.updateDummy` (formally `Caml_obj.update_dummy`) behavior is now the same as the `Object.assign` and doesn't specially handle `Array` instances. See discussion from #7043. But this is safer because it is only used in `module rec` bindings.
  - Removes the ascii bound restriction of char primitives.
  - `String.get: (string, int) => char` behavior is now the same as the `String.prototype.codePointAt`. It was potentially buggy with existing char primitives, but this fixes it. See #7028 
- Dropped built-in types
  - `int32` (use `int` instead)
  - `int64`
  - `uint32`
  - `nativeint`
  - `floatarray` (use `Float32Array` or `Float64Array` instead)
  - `bytes` (use `ArrayBuffer` or `Uinr8Array` instead)
- Renamed primitives
  - comparison
    - `caml_{type}_compare` -> `%{type}order` (e.g. `%intorder`)
    - `caml_{type}_min` -> `%{type}min` (e.g. `%intmin`)
    - `caml_{type}_max` -> `%{type}max` (e.g. `%intmax`)
    - `%bs_min` -> `%min`
    - `%bs_max` -> `%max`
    - `%bs_equal_null` -> `%equal_null`
    - `%bs_equal_undefined` -> `%equal_undefined`
    - `%bs_equal_nullable` -> `%equal_nullable`
    - `#unsafe_lt` -> `%unsafe_lt`
    - `#unsafe_gt` -> `%unsafe_gt`
    - `#unsafe_le` -> `%unsafe_le`
    - `#unsafe_ge` -> `%unsafe_ge`
    - `#unsafe_eq` -> `%unsafe_eq`
    - `#unsafe_neq` -> `%unsafe_neq`
    - `%bs_equal_null` -> `%equal_null`
    - `%bs_equal_undefined` -> `%equal_undefined`
    - `%bs_equal_nullable` -> `%equal_nullable`
  - js
    - `#debugger` -> `%debugger`
    - `#typeof` -> `%typeof`
    - `#null` -> `%null`
    - `#undefined` -> `%undefined`
  - ref
    - `%makemutable` -> `%makeref`
    - `%bs_field0` -> `%refget`
    - `%bs_set_field0` -> `%refset`
  - dict
    - `?create_dict` -> `%makedict`
  - list
    - `#makemutablelist` -> `%makemutablelist`
  - promise
    - `?await` -> `%await`
  - module
    - `#import` -> `%import`
  - curry
    - `#apply{N}` -> `%curry_apply{N}`
  - option
    - `#is_nullable` -> `%is_nullable`
    - `#nullable_to_opt` -> `%nullable_to_opt`
    - `#null_to_opt` -> `%null_to_opt`
    - `#undefined_to_opt` -> `%undefined_to_opt`
  - exceptions
    - `#wrap_exn` -> `%wrap_exn`
  - hash
    - `#hash` -> `%hash`
    - `#hash_mix_int` -> `%hash_mix_int`
    - `#hash_mix_string` -> `%hash_mix_string`
    - `#hash_final_mix` -> `%hash_final_mix`
  - etc
    - `#unsafe_to_method` -> `%unsafe_to_method`
    - `#function_length` -> `%function_arity`
- Dropped primitives
  - `?obj_dup` (from #7043)
  - formatting
    - `?hexstring_of_float`
    - `?float_of_string`
    - `?int_of_string`
    - `?int64_of_string`
    - `?int64_format`
    - `?format_int`
    - `?format_float`
  - parsing
    - `?lex_engine`
    - `?new_lex_engine`
    - `?parse_engine`
    - `?set_parser_trace`
  - hashing
    - `?md5_string`
  - string
    - `?string_repeat`
  - bytes
    - `?create_bytes`
    - `?bytes_length`
    - `?bytes_safe_get`
    - `?bytes_safe_set`
    - `?bytes_unsafe_get`
    - `?bytes_unsafe_set`
    - `caml_bytes_*`
  - int64
    - `%int64_*`
    - `caml_int64_*`
  - bigint
    - `?bigint_div`
    - `?bigint_mod`
    - `%bigint_of_int32`
    - `%bigint_to_int32`
  - float
    - `?int_of_float` (`%intoffloat` is still available)
    - `?float_of_int` (`%floatofint` is still available)
    - `?int_float_of_bits`
    - `?int_bits_of_float`
    - `?fmod_float`
    - `?modf_float`
    - `?ldexp_float`
    - `?frexp_float`
    - `?copysign_float`
    - `?expm1_float`
    - `?hypot_float`
  - exn
    - `?exn_slot_name`
    - `?is_extension`
    - `?as_js_exn`
    - `%reraise`
    - `%raise_notrace`
  - lazy
    - `%lazy_force` (Use `Lazy.force` instead)
- Dropped BS compile-time constants
  - `%backend_type`
  - `%word_size`
  - `%int_size`
  - `%max_wosize`
- Dropped OCaml literal
  - `0u` (uint32) literal
  - `0l` (int32) literal
  - `0L` (int64) literal
- Dropped Js functions (which was not Js)
  - `Js.isCamlExceptionOrOpenVariant`
- Dropped pervasive functions
  - `format_float`
  - `string_of_bool`
  - `bool_of_string`
  - `bool_of_string_opt`
  - `string_of_int`
  - `int_of_string`
  - `int_of_string_opt`
  - `string_of_float`
  - `float_of_string`
  - `float_of_string_opt`
  - `print_string`
  - `print_int`
  - `print_float`
  - `print_endline`
  - `print_newline`
  - `prerr_endline`
  - `prerr_newline`
  - `valid_float_lexem`
- Dropped OCaml std modules
  - `Arg`
  - `Callback`
  - `Genlex`
  - `Lexing`
  - `Parsing`
  - `Parsing`
  - `Digest`
  - `Random`
  - `Bytes`
  - `Uchar`
  - `Buffer`
  - `Stream`
  - `Filename`
  - `Int32`
  - `Int64`
  - `String` partially
  - `Char` partially
  - `Array`
  - `Sort`
  - `List`
  - `Set`
  - `Map`
  - `Hashtbl` (except `Hashtbl.hash`)
  - `Map`
  - `Stack`
  - `Queue`
  - `Complex`
- Modified primitive runtime modules
  - `Caml_module` -> `Primitive_module`
  - `Caml_obj` -> `Primitive_object`
  - `Caml_array` -> `Primitive_array`
  - `Caml_char` -> `Primitive_char`
  - `Caml_string` -> `Primitive_string`
  - `Caml_exceptions`, `Caml_js_exceptions` -> `Primitive_exceptions`
  - `Caml_hash`, `Caml_hash_primitive` -> `Primitive_hash`
  - `Caml_lazy` -> `Primitive_lazy`
  - `Caml_module` -> `Primitive_module`
  - `Caml_option` -> `Primitive_option`
  - `Curry` -> `Primitive_curry`
  - `Runtime_promise` -> `Primitive_promise`
  - `Runtime_dict` -> `Primitive_dict`
  - `Runtime_deriving` -> `Primitive_util`
  - x -> `Primitive_bool`
  - x -> `Primitive_int`
  - x -> `Primitive_float`
  - x -> `Primitive_bigint`
  - `Caml_format` -> x
  - `Caml_sys` -> x
  - `Caml_lexer` -> x
  - `Caml_parser` -> x
  - `Caml_md5` -> x
  - `Caml_int32` -> x
  - `Caml_int64` -> x
  - `Caml_splice_call` -> x (from #7030)

## Note for future refactoring

All primitives except # prefixed should eventually be treated as "features" and documented. We can rename all by a new convention for primitive names, but the old names should be kept for a while for compatibility.

Primitives are declared in two different places because they are processed separately before and after `Lam_convert`. It is difficult to change and to determine the types used or not used, resulting in many dangerous assertions.

Ultimately, I want to consolidate the primitive declarations into one place and make them easier to manage.

DZakh commented:
Like it

cometkim commented:
Maybe it's better to merge #6979 first, then I'll rebase

cristianoc commented:
> Maybe it's better to merge #6979 first, then I'll rebase

Definitely.
Given the size of this amazing work we can't really review it but we can quickly observe if all the generated code is simply identical.

cometkim commented:
Actually, it was a bit improved like https://github.com/rescript-lang/rescript-compiler/pull/6984/commits/03fe31086dccfe98ca91fe66e92983df95bb26b6#diff-ef282bde01036b4b8ce35133e81733df0530c2e4192992862d2ef200c114e5e0

Having dedicated variants rather than just Pccall makes analyzing side effects /specialization much easier.

cometkim commented on jscomp/test/obj_magic_test.js:
```diff
@@ -22,10 +22,10 @@ let suites_1 = {
       TAG: "Eq",
       _0: true,
-      _1: typeof ({
+      _1: is_block({
```
de-opted, but better result...? I think we can fold this into a single bool value here.

cknitt commented:
I tested against one of our projects and found that the array indexing now does not return `option` anymore, even if Core is open.

I.e. this compiles:

```rescript
open RescriptCore

let a = [1, 2, 3]

let x: int = a[0]
let y: option<int> = a->Array.get(0)
```

cknitt commented:
> Maybe it's better to merge https://github.com/rescript-lang/rescript-compiler/pull/6979 first, then I'll rebase

Not sure as #6979 is currently blocked by the binary incompatibility issue described in https://github.com/rescript-lang/rescript-compiler/pull/6979/files#r1738735381, so we would need to wait for that to be resolved.

zth commented:
> > Maybe it's better to merge #6979 first, then I'll rebase
> 
> Not sure as #6979 is currently blocked by the binary incompatibility issue described in https://github.com/rescript-lang/rescript-compiler/pull/6979/files#r1738735381, so we would need to wait for that to be resolved.

I'm going to see if we can figure out a quick fix ASAP for just this situation so that can get merged, so that and this can move along. Will get back here when I got something.

zth commented:
> > > Maybe it's better to merge #6979 first, then I'll rebase
> > 
> > 
> > Not sure as #6979 is currently blocked by the binary incompatibility issue described in https://github.com/rescript-lang/rescript-compiler/pull/6979/files#r1738735381, so we would need to wait for that to be resolved.
> 
> I'm going to see if we can figure out a quick fix ASAP for just this situation so that can get merged, so that and this can move along. Will get back here when I got something.

Actually seems like the binary compat was already broken here: https://github.com/rescript-lang/rescript-compiler/pull/6943

Me and @cristianoc discussed a bit, and we think it's fine to merge that PR as is, because the editor extension doesn't touch any of these things where we've extended the AST. However, we need to figure out a real solution for this mid to long term. 

The real solution is likely to be that the extension ships its own compiler libs vendored that's compatible with `<=11.x`, and we start shipping the language server runnable via the compiler for v12+ going forward, so that changes made to the AST is fine to make. 

However, that's going to be some work, and I don't have time for that right now. The hard "deadline" for this work is when we introduce changes that break the binary compat in ways that matter for the editor extension. So, it'll be fine for now.

cometkim commented:
> I tested against one of our projects and found that the array indexing now does not return `option` anymore, even if Core is open.
> 
> I.e. this compiles:
> 
> ```rescript
> open RescriptCore
> 
> let a = [1, 2, 3]
> 
> let x: int = a[0]
> let y: option<int> = a->Array.get(0)
> ```

Fixed in https://github.com/rescript-lang/rescript-compiler/pull/6984/commits/890fb4a0813ce8c732f9a2d8e585a8a77811ecd9

This is related to the behavior I mentioned in Discord. It's quite confusing that `open MyStd` or `module Array = ...` can break types of primitive syntax.

cometkim commented:
> I tested against one of our projects and found that the array indexing now does not return `option` anymore, even if Core is open.
> 
> I.e. this compiles:
> 
> ```rescript
> open RescriptCore
> 
> let a = [1, 2, 3]
> 
> let x: int = a[0]
> let y: option<int> = a->Array.get(0)
> ```

Fixed in https://github.com/rescript-lang/rescript-compiler/pull/6984/commits/890fb4a0813ce8c732f9a2d8e585a8a77811ecd9

This is related to the behavior I mentioned in Discord. It's quite confusing that `open MyStd` or `module Array = ...` can break types of primitive syntax.

cknitt commented:
I tested again against one of our projects. I ran into some things like `mod_float` and `string_of_int` missing and had to patch some dependencies, but once I got it to build, there were no diffs in the output compared to alpha.3! 🎉

So I guess we should keep at least the more widely used functions like `mod_float` and `string_of_int`, but mark them as deprecated in v12.

cometkim commented:
Added compatibility for Pervasives, Array, List, Char, and String.

Additionally, I modified the `ninja.js` to make `runtime` dir include only primitives and `others` dir include only stdlib. `Pervasives` now belong solely to primitives, and `Js` should be part of the stdlib. No-more copy-pasting and ordering hacks

The dependency graph is now `others` -> `runtime` clearly

cknitt commented:
`string_of_int` and `mod_float` are still missing.

For the deprecations, I would suggest to only have "Use Core instead", not "Use Core or Js instead", as Js is also legacy.

Hmm, actually, once Core is in the compiler, we should refer to the matching function without mentioning Core at all, e.g., "Use `Float.toString` instead."

zth commented:
Core will not be optional, as in it will always be available and on by default. This is because one of the big goals with Core is to provide a great default, and we don't want to give the impression that it's "Core or these other available things" . Core won't be a separate concept anymore when it's integrated in the compiler, it'll just be a part of the language. 

However, for advanced users it'll work exactly like today - you can shadow it easily with your own stdlib if you want, just like you can shadow the OCaml and/or the Js stuff today.

zth commented:
@cometkim IMO you don't need to spend time on getting the error messages right here, we'll have to adjust them anyway as Core is brought in. 

Yeah, that was a somewhat calculated breaking change. The positive side is that anyone who's moved to Core already (which I hope is the majority) will already have dealt with that.

Anyway, again, absolutely fantastic job on this PR. A pretty insane feat changing this amount of code and causing no major breakages!

cknitt commented on Makefile:
```diff
@@ -84,7 +84,6 @@ clean-gentype:
 
 clean:
 	dune clean
-	cargo clean --manifest-path rewatch/Cargo.toml && rm -f rewatch/rewatch
```
I would prefer to keep this. `make clean` should clean everything including the rewatch build.

cometkim replied:
then what's different with `make clean-all`?

It's inconvenient to rebuild `rewatch` every time I run `make clean test`. we can add `make clean-rewatch`

cknitt replied:
Ok, this is a good question. 😄 This Makefile has grown organically.

I guess it's also fine to have `clean-rewatch` and put that into `clean-all`, too.

cknitt commented on jscomp/others/js_exn.resi:
```diff
@@ -31,16 +31,13 @@ type t
 type exn += private Error(t)
 
-external asJsExn: exn => option<t> = "?as_js_exn"
+let asJsExn: exn => option<t>
```
This means it won't be zero cost anymore?

cometkim replied:
It has runtime in primitive anyway

cknitt replied:
Ok! 👍

cknitt commented:
@cometkim Skimmed through the rest of the changes. Looks good to me! 👍

Could you just bring back the "cargo clean" in the Makefile in some form or other a discussed above? Then good to go from my point of view.

cometkim commented:
> Could you just bring back the "cargo clean" in the Makefile in some form or other a discussed above? Then good to go from my point of view.

Added `clean-rewatch`, and rebased

cknitt commented:
Merged! 🎉
